File: api/code/api.integration.tests/Cosmos.cs
using Aspire.Microsoft.Azure.Cosmos;
using Azure.Core;
using common;
using LanguageExt;
using Microsoft.Azure.Cosmos;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Hosting;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace api.integration.tests;

internal delegate ValueTask EmptyOrdersContainer(CancellationToken cancellationToken);

internal sealed record OrdersContainer(Container Value);

internal static class CosmosModule
{
    public static void ConfigureEmptyOrdersContainer(IHostApplicationBuilder builder)
    {
        ConfigureOrdersContainer(builder);

        builder.Services.TryAddSingleton(GetEmptyOrdersContainer);
    }

    private static EmptyOrdersContainer GetEmptyOrdersContainer(IServiceProvider provider)
    {
        var ordersContainer = provider.GetRequiredService<OrdersContainer>();
        var activitySource = provider.GetRequiredService<ActivitySource>();

        var container = ordersContainer.Value;

        return async cancellationToken =>
        {
            using var activity = activitySource.StartActivity(nameof(EmptyOrdersContainer));

            var deletedCount = 0;

            await listOrders(cancellationToken)
                    .IterParallel(async x =>
                    {
                        var (cosmosId, partitionKey, eTag) = x;

                        await common.CosmosModule.DeleteRecord(container,
                                                               cosmosId,
                                                               partitionKey,
                                                               eTag,
                                                               cancellationToken);

                        Interlocked.Increment(ref deletedCount);
                    }, cancellationToken);

            activity?.AddTag(nameof(deletedCount), deletedCount);
        };

        IAsyncEnumerable<(CosmosId, PartitionKey, ETag)> listOrders(CancellationToken cancellationToken)
        {
            var query = new CosmosQueryOptions
            {
                Query = new QueryDefinition("""
                    SELECT c.id, c.orderId, c._etag
                    FROM c
                    """)
            };

            return common.CosmosModule.GetQueryResults(container, query, cancellationToken)
                         .Select(json => from id in common.CosmosModule.GetCosmosId(json)
                                         from partitionKeyString in json.GetStringProperty("orderId")
                                         let partitionKey = new PartitionKey(partitionKeyString)
                                         from eTag in common.CosmosModule.GetETag(json)
                                         select (id, partitionKey, eTag))
                         .Select(result => result.ThrowIfFail());
        }
    }

    private static void ConfigureOrdersContainer(IHostApplicationBuilder builder)
    {
        ConfigureDatabase(builder);

        builder.Services.TryAddSingleton(GetOrdersContainer);
    }

    private static OrdersContainer GetOrdersContainer(IServiceProvider provider)
    {
        var database = provider.GetRequiredService<Database>();
        var configuration = provider.GetRequiredService<IConfiguration>();

        var containerName = configuration.GetValueOrThrow("COSMOS_ORDERS_CONTAINER_NAME");

        return new(database.GetContainer(containerName));
    }

    private static void ConfigureDatabase(IHostApplicationBuilder builder)
    {
        ConfigureCosmosClient(builder);

        builder.Services.TryAddSingleton(GetDatabase);
    }

    private static Database GetDatabase(IServiceProvider provider)
    {
        var client = provider.GetRequiredService<CosmosClient>();
        var configuration = provider.GetRequiredService<IConfiguration>();

        var databaseName = configuration.GetValueOrThrow("COSMOS_DATABASE_NAME");

        return client.GetDatabase(databaseName);
    }

    private static void ConfigureCosmosClient(IHostApplicationBuilder builder)
    {
        AzureModule.ConfigureTokenCredential(builder);

        var configuration = builder.Configuration;
        var connectionName = configuration.GetValue("COSMOS_CONNECTION_NAME").IfNone(string.Empty);
        var provider = builder.Services.BuildServiceProvider();

        builder.AddAzureCosmosClient(connectionName, configureSettings, configureClientOptions);

        void configureSettings(MicrosoftAzureCosmosSettings settings)
        {
            configuration
                .GetValue("COSMOS_ACCOUNT_ENDPOINT")
                .Map(endpoint => new Uri(endpoint, UriKind.Absolute))
                .Iter(endpoint =>
                {
                    settings.AccountEndpoint = endpoint;
                    settings.Credential = provider.GetRequiredService<TokenCredential>();
                });

            configuration
                .GetValue("COSMOS_CONNECTION_STRING")
                .Iter(connectionString => settings.ConnectionString = connectionString);
        }

        void configureClientOptions(CosmosClientOptions options)
        {
            options.EnableContentResponseOnWrite = false;
            options.UseSystemTextJsonSerializerWithOptions = JsonSerializerOptions.Web;
            options.CosmosClientTelemetryOptions = new() { DisableDistributedTracing = false };
        }
    }
}

File: api/code/api.integration.tests/Cosmos.fs
namespace api.integration.tests

open Aspire.Microsoft.Azure.Cosmos
open Azure.Core
open Microsoft.Azure.Cosmos
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.DependencyInjection
open System
open System.Text.Json
open FSharpPlus
open FSharp.Control
open common

type OrdersContainer =
    | OrdersContainer of Container

    static member toContainer(OrdersContainer container) = container

[<RequireQualifiedAccess>]
module private Cosmos =
    let private deleteOrder (OrdersContainer container) json =
        let cosmosId, partitionKey, eTag =
            monad {
                let! id = Cosmos.getId json
                let! partitionKey = PartitionKey.fromOrderJson json
                let! eTag = Cosmos.getETag json
                return id, partitionKey, eTag
            }
            |> JsonResult.throwIfFail

        async {
            match! Cosmos.deleteRecord container partitionKey cosmosId eTag with
            | Ok() -> return ()
            | Error error -> return failwith $"Failed to delete order. Error is '{error}'."
        }


    let emptyContainer ordersContainer =
        async {
            let query =
                CosmosQueryOptions.fromQueryString "SELECT c.id, c.orderId, c._etag FROM c"

            let container = OrdersContainer.toContainer ordersContainer

            do!
                Cosmos.getQueryResults container query
                |> AsyncSeq.iterAsyncParallel (deleteOrder ordersContainer)
        }

    let private configureCosmosSettings configuration provider (settings: MicrosoftAzureCosmosSettings) =
        let getConfigurationValue = Configuration.getValue configuration

        let tryParseUri value =
            match Uri.TryCreate(value, UriKind.Absolute) with
            | true, uri -> Some uri
            | _ -> None

        getConfigurationValue "COSMOS_ACCOUNT_ENDPOINT"
        |> bind tryParseUri
        |> iter (fun endpoint ->
            settings.AccountEndpoint <- endpoint
            let tokenCredential = ServiceProvider.getServiceOrThrow<TokenCredential> provider
            settings.Credential <- tokenCredential)

        getConfigurationValue "COSMOS_CONNECTION_STRING"
        |> Option.iter (fun connectionString -> settings.ConnectionString <- connectionString)

    let private configureClientOptions (options: CosmosClientOptions) =
        options.EnableContentResponseOnWrite <- false
        options.UseSystemTextJsonSerializerWithOptions <- JsonSerializerOptions.Web

        options.CosmosClientTelemetryOptions <-
            let options = CosmosClientTelemetryOptions()
            options.DisableDistributedTracing <- false
            options

    let private configureClient (builder: IHostApplicationBuilder) =
        Azure.configureTokenCredential builder

        let configuration = builder.Configuration

        let connectionName =
            Configuration.getValue configuration "COSMOS_CONNECTION_NAME"
            |> Option.defaultValue String.Empty

        let configureCosmosSettings =
            configureCosmosSettings configuration (builder.Services.BuildServiceProvider())

        builder.AddAzureCosmosClient(connectionName, configureCosmosSettings, configureClientOptions)

    let private getDatabase (provider: IServiceProvider) =
        let client = provider.GetRequiredService<CosmosClient>()
        let configuration = provider.GetRequiredService<IConfiguration>()

        let databaseName =
            Configuration.getValueOrThrow configuration "COSMOS_DATABASE_NAME"

        client.GetDatabase(databaseName)

    let private configureDatabase builder =
        configureClient builder

        ServiceCollection.tryAddSingleton builder.Services getDatabase

    let private getOrdersContainer (provider: IServiceProvider) =
        let database = provider.GetRequiredService<Database>()
        let configuration = provider.GetRequiredService<IConfiguration>()

        let containerName =
            Configuration.getValueOrThrow configuration "COSMOS_ORDERS_CONTAINER_NAME"

        database.GetContainer(containerName) |> OrdersContainer

    let configureOrdersContainerBuilder builder =
        configureDatabase builder

        ServiceCollection.tryAddSingleton builder.Services getOrdersContainer
